### DISCUSSION TITLE
screen <--> world using depth buffer & question about world crs

### DISCUSSION THREAD

**Discussion by rebmaso** [pelicanmapping/osgearth#2366] [Support Q&A] [answered]
Hi everyone,

I'm trying to get world coordinates from screen coordinates.
What I'm seeing is that **neither world --> screen nor screen --> world are working in my case**.
I believe there's an **issue in what I consider world coordinates**. 

Here's my full repo for reference. You can look at src/main.cpp. 

https://github.com/rebmaso/osgearth_renderer

### World to Screen

I did my homework and came across issue https://github.com/gwaldron/osgearth/issues/1196 , where this snippet is provided by @gwaldron:

```c++
osg::Matrix VPW = viewMatrix * projectionMatrix * windowMatrix;
osg::Vec3d screenCoord = worldCoords * VPW;
```

Now, **I'm assuming OsgEarth uses the ECEF** world reference frame. So I'm specifying points in ECEF and projecting them with the formula.
My issue is that **projected points according to this formula are not where I actually see them**, but nearby (not completely out of the frame). **Should I use a different reference system to express worldCoords**?

### Screen to World

I'm encountering the same issue doing the opposite thing: the world coordinates are skewed, but still somewhat near the right coordinates.

I know there is `bool
EarthManipulator::screenToWorld(float x, float y, osg::View* theView, osg::Vec3d& out_coords) const` to do this, but I'd like to avoid `Earthmanipulator`, using instead the depth buffer & the inverse VPW matrix, as also explained in the OpenSceneGraph Beginnners' guide, p. 164 ("From world to screen").

This is a snip of my code:

```c++

// Allocate image with depth buffer
osg::ref_ptr<osg::Camera> osgCam = viewer.getCamera();
osg::ref_ptr<osg::Image> zImage = new osg::Image;
osgCam->attach(osg::Camera::DEPTH_BUFFER, zImage.get());
zImage->allocateImage(MY_OSG_WIDTH, MY_OSG_HEIGHT, 1, GL_DEPTH_COMPONENT , GL_FLOAT);

osg::Matrixf VPW = viewer.getCamera()->getViewMatrix() *
            viewer.getCamera()->getProjectionMatrix() * 
             viewer.getCamera()->getViewport()->computeWindowMatrix();

osg::Matrix invVPW;

if(!invVPW.invert(VPW)) LOG(FATAL) << "matrix inversion failed!";

float screenCoords[2] = {200, 200}; // xy

// osg::Vec3f nearPoint = osg::Vec3f(screenCoords[0], screenCoords[1], 0.f) * invVPW;
// osg::Vec3f farPoint = osg::Vec3f(screenCoords[0], screenCoords[1], 1.f) * invVPW;
float z_buffer = ((float*)zImage->data(screenCoords[0], screenCoords[1]))[0];

// hom to eucl handled implicitly in the overloaded *
osg::Vec3f worldCoords = osg::Vec3f(screenCoords[0], screenCoords[1], z_buffer) * invVPW;

std::cout << "z_buffer: " << z_buffer << "\n";
std::cout << "worldCoords: " << worldCoords[0] << "  " << worldCoords[1] << "  " << worldCoords[2] << "\n";

```

**Comment by gwaldron** [comment]
If it's close but not exact, first thing I would do it change that `osg::Vec3f` to `osg::Vec3d`. All those matrices are double-precision so you need to use a double-precision coordinate.

(I assume you are actually putting your depth-buffer camera in the scene graph and letting it run before doing this right? Just asking since you don't show it in the snippet)

**Reply by rebmaso** [reply]
Thanks for replying so fast, Glenn.

1) I'm now using double precision. Nothing changes.

2) Right now I'm just manually calling `viewer.frame()` a bunch of times after attaching `zImage` to the camera, like this:

``` c++
osg::ref_ptr<osg::Camera> osgCam = viewer.getCamera();
osg::ref_ptr<osg::Image> zImage = new osg::Image;
osgCam->attach(osg::Camera::DEPTH_BUFFER, zImage.get());
zImage->allocateImage(MY_OSG_WIDTH, MY_OSG_HEIGHT, 1, GL_DEPTH_COMPONENT , GL_FLOAT);
for(size_t i = 0; i < 100; i++) viewer.frame();
```

I saved the depth buffer image as a tif and it looks OK. The values I read are between 0.0 and 1.0. 

Maybe I should start by  **figuring out why the simple world to screen case is not working**. 

With that respect, this is my test case. Do you see any inconsistencies? You can see that `screen` actually has a depth component. Does that look right? Do I just take the first two elements of `screen` as my pixel coordinates?

``` c++
osg::Vec3d ecef = osg::Vec3d(4367869.861102294, 502992.54553608876, 4605567.651819189);
osg::Vec3d screen = ecef * VPW;

std::cout << "screen: " << screen[0] << " " << screen[1] << " " << screen[2] << "\n";
```

Output:
```
screen: 457.201 121.272 0.332592
```

Thanks

**Comment by timoore** [comment]
Logarithmic depth buffer?

**Comment by rebmaso** [comment]
@timoore https://github.com/gwaldron/osgearth/discussions/2366#discussioncomment-7372430 I'm kind of new to computer graphics. It should be logarithmic right?

Anyway, This is my repo with full code in case anyone wants to take a peek at src/main.cpp (messy, but few lines of code).

https://github.com/rebmaso/osgearth_renderer

**Comment by timoore** [comment]
> @timoore [#2366 (comment)](https://github.com/gwaldron/osgearth/discussions/2366#discussioncomment-7372430) I'm kind of new to computer graphics. It should be logarithmic right?

No, I mean that osgearth uses a logarithmic representation of the non-linear depth values written by OpenGL. This is one strategy to overcome precision issues and Z fighting. I'm not clear on whether or not the logarithmic depth buffer is on by default. If it is, then you need to use a modification of the classic formula, which I'm sure that @gwaldron will be happy to provide :smile: 
> 
> Anyway, This is my repo with full code in case anyone wants to take a peek at src/main.cpp (messy, but few lines of code).
> 
> https://github.com/rebmaso/osgearth_renderer

**Reply by gwaldron** [reply]
@timoore theory is easy enough to test -- run your application with this command-line argument:
`--nologdepth`

**Comment by rebmaso** [comment]
Hi all.
Thanks for the great input. I'm getting closer to a working solution and **will provide feedback and clean functioning code ASAP** for anybody who stumbles into this issue in the future.
Thanks so much!

**Comment by rebmaso** [comment]
[Update]

I tried the `--nologdepth` flag to use the standard nonlinear depth buffer. World coordinates are still not correct.

**Weird**: I also tried using the `EarthManipulator::screenToWorld(float x, float y, osg::View* theView, osg::Vec3d& out_coords)` method. By taking a point at openGL (origin bottom left) image coordinates x = 200, y = 200, I get wrong world coordinates! **However**, if I launch the app using the `--coords` flag and hover over said image point, the geodetic world coordinates that are displayed are correct.

Am I passing wrong X Y image coordinates (both using the VPV matrix and the osgearth screenToWorld method)?

As @gwaldron suggested in thread https://github.com/gwaldron/osgearth/discussions/2374 , I will also try to use the logarithmic depth buffer and convert log to nonlinear z values before applying the invMVPV matrix.

``` c++

osg::ref_ptr<osg::Camera> osgCam = viewer.getCamera();

osg::ref_ptr<osg::Image> zImage = new osg::Image;


zImage->allocateImage(MY_OSG_WIDTH, MY_OSG_HEIGHT, 1, GL_DEPTH_COMPONENT , GL_FLOAT);
osgCam->attach(osg::Camera::DEPTH_BUFFER, zImage.get());

osg::Matrixd VPW = viewer.getCamera()->getViewMatrix() *
          viewer.getCamera()->getProjectionMatrix() * 
           viewer.getCamera()->getViewport()->computeWindowMatrix();

  osg::Matrixd invVPW;

  if(!invVPW.invert(VPW)) {
      std::cerr << "[ERROR] Matrix inversion failed.";
      exit(1);
  }

// ========= Test world <--> screen =========

  // in opengl, screen origin is bottom left
  float screenCoords[2] = {200., 200.}; // xy

  // access by column row
  double zBuffer = ((double*)zImage->data(screenCoords[0], screenCoords[1]))[0];

  // can also use vec3, hom to eucl handled implicitly in the overloaded *
  osg::Vec4d worldCoords = osg::Vec4d(screenCoords[0], screenCoords[1], zBuffer, 1.) * invVPW;
  
  std::cout << "========= Test depth -> world ========= \n";
  fmt::print("screenCoords: {} {} \n" , screenCoords[0] , screenCoords[1]);
  fmt::print("zBuffer: {} \n", zBuffer);
  fmt::print("worldCoords via depth: {}, {} ,{} \n" ,  worldCoords[0]/worldCoords[3] , worldCoords[1]/worldCoords[3] , worldCoords[2]/worldCoords[3] );
  // try with osgearth 
  osg::Vec3d out_coords;
  // same as mapNode->getTerrain()->getWorldCoordsUnderMouse(view, x, y, out_coords);
  if(!manip->screenToWorld(screenCoords[0], screenCoords[1], &viewer, out_coords))
      std::cerr << "Screen to world failed! \n";
  else 
      fmt::print("worldCoords via osgearth: {}, {} ,{} \n" , out_coords[0] , out_coords[1] , out_coords[2]);


  // ====================================
```

**Comment by gwaldron** [comment]
Also don't forget that the Z-buffer will be in the range [0..1] (by default) while OpenGL clip coordinates are z=[-1..1].

**Reply by rebmaso** [reply, marked as answer]
I managed to get a working solution.

First, I run my viewer app passing the --nologdepth flag (suggested by Tim and Glenn!) from command line (perhaps there's a way to pass it to osgViewer inside the code). 

Then, I take a deep copy of the depth buffer and the InvVPW matrix using the following functions:

``` c++

void OsgInterface::getScreenShot() {

    if (!initialized) {
            fmt::print("Waiting for the scene to load... \n\n");
            // Wait for scene to load for the first time
            for(size_t i = 0; i< first_delay_frames_; i++) viewer->frame();
            // Set the flag to indicate initialization has occurred
            initialized = true;
        }
    else {
    // Wait for render to be ready - otherwise black image!
    for(size_t i = 0; i < screenshot_delay_frames_ ; i++) viewer->frame();
    }

    if(!zImage->isDataContiguous()) {
        std::cerr << "[ERROR] depth data not contiguous!";
        exit(1);
    }

    if(!colorImage->isDataContiguous()) {
        std::cerr << "[ERROR] color data not contiguous!";
        exit(1);
    }

    cv::Mat tempDepthMatOut(zImage->t(), zImage->s(), CV_32FC1);
    tempDepthMatOut.data = (uchar*)zImage->data();
    depthMat = tempDepthMatOut.clone();
    cv::flip(depthMat, depthMat, 0);

    cv::Mat tempRefImageMatOut(colorImage->t(), colorImage->s(), CV_8UC3);
    tempRefImageMatOut.data = (uchar*)colorImage->data();
    refImageMat = tempRefImageMatOut.clone();
    cv::flip(refImageMat, refImageMat, 0);

}


void OsgInterface::getInvVPW() {

    // Ref: OpenSceneGraph beginnners guide, p 164

    // screen = world * VPW 
    // world = screen * VPW_inv

    // use MVPW if world coordinates are in model reference frame

    osg::Matrixd VPW = viewer->getCamera()->getViewMatrix() *
            viewer->getCamera()->getProjectionMatrix() * 
             viewer->getCamera()->getViewport()->computeWindowMatrix();

    osg::Matrixd VP = viewer->getCamera()->getViewMatrix() *
            viewer->getCamera()->getProjectionMatrix();

    if(!invVPW.invert(VPW)) {
        std::cerr << "[ERROR] Matrix inversion failed.";
        exit(1);
    }

    Eigen::Map<const Eigen::Matrix<double, 4, 4, Eigen::RowMajor>> tempInvVPWEig((double*)invVPW.ptr());

    invVPWEig = tempInvVPWEig; // deep copy

}
```

At this point, ECEF coordinates of image points can be accessed with the following helper function:

``` c++
void depthToCoordsOSG(const cv::Mat & depthMat,
                const Eigen::MatrixXd & invVPWMat,
                const std::vector<cv::Point2f> & refPoints,
                const std::vector<cv::Point2f> & queryPoints,
                std::vector<cv::Point2d> & points2DGood,
                std::vector<cv::Point3d> & points3DGood
                ) {

    std::vector<float> x_d, y_d;
    for (const auto& point : refPoints) {
        x_d.push_back(point.x);
        y_d.push_back(point.y);
    }

    // Get openscenegraph viewport coords (cols(x), HEIGHT - rows(y))
    // Keypoints are x-y
    Eigen::VectorXf x_gl = Eigen::Map<const Eigen::VectorXf>(x_d.data(), x_d.size());
    Eigen::VectorXf y_gl = depthMat.rows - Eigen::Map<const Eigen::VectorXf>(y_d.data(), y_d.size()).array();
    
    // fill depth
    Eigen::VectorXf z_gl(y_d.size());
    for (size_t i = 0; i < y_d.size(); ++i) {
        z_gl(i) = depthMat.at<float>(static_cast<int>(y_d[i]), static_cast<int>(x_d[i]));
    }

    Eigen::VectorXf w_gl = Eigen::VectorXf::Ones(y_d.size());

    // Combine into a matrix
    Eigen::MatrixXf viewportCoords(4, x_gl.size());
    viewportCoords.row(0) = x_gl.transpose();
    viewportCoords.row(1) = y_gl.transpose();
    viewportCoords.row(2) = z_gl.transpose();
    viewportCoords.row(3) = w_gl.transpose();

    // Get ECEF coords by converting to world using invVPW
    Eigen::MatrixXd points3D = viewportCoords.transpose().cast<double>() * invVPWMat;

    // Filter out 2D and 3D points with invalid depth
    // this is 1 where condition is met, 0 where not met
    // TODO: this should be done before computations, but easier here!
    Eigen::Array<bool, Eigen::Dynamic, 1> goodDepthIndices = (z_gl.array() != 0.0f) && (z_gl.array() != 1.0f);
    for (long int i = 0; i < goodDepthIndices.size(); ++i) {
        if (goodDepthIndices[i]) {
            // 3D points - convert hom to euclid
            cv::Point3d point3D(points3D(i, 0) / points3D(i, 3), 
                                points3D(i, 1) / points3D(i, 3), 
                                points3D(i, 2) / points3D(i, 3));
            points3DGood.push_back(point3D);
            // 2D points
            points2DGood.push_back(queryPoints[i]);
        }
    }

    // std::cout << " 3D & 2d sizes: " << points3DGood.size() << " " << points2DGood.size() << " \n"; 
}
```
